Source/Urho3D/Network/HttpRequest.cpp

```diff
@@ -22,94 +22,95 @@
 
 #include "../Precompiled.h"
 
-#include "../Core/Profiler.h"
-#include "../Core/Timer.h"
-#include "../IO/Log.h"
-#include "../Network/HttpRequest.h"
+#include <Urho3D/Core/Profiler.h>
+#include <Urho3D/Core/Timer.h>
+#include <Urho3D/IO/Log.h>
+#include <Urho3D/Network/HttpRequest.h>
 
+#ifdef URHO3D_PLATFORM_WEB
+#include <emscripten.h>
+#else
 #include <Civetweb/civetweb.h>
+#endif
 
 #include "../DebugNew.h"
 
 namespace Urho3D
 {
 
 static const unsigned ERROR_BUFFER_SIZE = 256;
-static const unsigned READ_BUFFER_SIZE = 65536; // Must be a power of two
+static const unsigned READ_BUFFER_SIZE = 1024;
 
 HttpRequest::HttpRequest(const ea::string& url, const ea::string& verb, const ea::vector<ea::string>& headers, const ea::string& postData) :
     url_(url.trimmed()),
     verb_(!verb.empty() ? verb : "GET"),
     headers_(headers),
-    postData_(postData),
-    state_(HTTP_INITIALIZING),
-    httpReadBuffer_(new unsigned char[READ_BUFFER_SIZE]),
-    readBuffer_(new unsigned char[READ_BUFFER_SIZE]),
-    readPosition_(0),
-    writePosition_(0)
+    postData_(postData)
 {
     // Size of response is unknown, so just set maximum value. The position will also be changed
     // to maximum value once the request is done, signaling end for Deserializer::IsEof().
     size_ = M_MAX_UNSIGNED;
+    URHO3D_LOGDEBUG("HTTP {} request to URL {}", verb_, url_.ToString());
 
-    URHO3D_LOGDEBUG("HTTP " + verb_ + " request to URL " + url_);
+#if defined(URHO3D_PLATFORM_WEB)
+    if (!headers.empty())
+        Log::GetLogger("HttpRequest").Warning("Custom headers in Web builds are not supported and were ignored!");
 
-#ifdef URHO3D_SSL
+    requestHandle_ = emscripten_async_wget2_data(url_.ToString().c_str(), verb_.c_str(), postData.c_str(), this, 1,
+        [](unsigned handle, void* arg, void* buf, unsigned len)         // onload
+        {
+            HttpRequest* request = static_cast<HttpRequest*>(arg);
+            MutexLock lock(request->mutex_);
+            request->state_ = HTTP_CLOSED;
+            request->readBuffer_.Resize(len);
+            request->readBuffer_.SetData(buf, len);
+            request->requestHandle_ = 0;
+        }, [](unsigned handle, void* arg, int code, const char* msg)    // onerror
+        {
+            HttpRequest* request = static_cast<HttpRequest*>(arg);
+            MutexLock lock(request->mutex_);
+            request->state_ = HTTP_ERROR;
+            request->error_ = msg;
+            request->requestHandle_ = 0;
+        }, [](unsigned handle, void* arg, int loaded, int total)        // onprogress
+        {
+            //HttpRequest* request = static_cast<HttpRequest*>(arg);
```

Comment:
That feels incomplete

Reply:
We need this lambda anyway as it is not optional, but our `HttpRequest` does not report progress so there is nothing to implement. Commented out line is there as a hint for how to access `HttpRequest` object in case we ever need to implement this callback.